bugsnag-android-core/src/main/java/com/bugsnag/android/Client.java

```diff
@@ -826,16 +798,18 @@ private void leaveErrorBreadcrumb(@NonNull Event event) {
     }
 
     @NonNull
-    private String getKeyFromClientData(Map<String, Object> clientData,
-                                        String key,
-                                        boolean required) {
-        Object value = clientData.get(key);
-        if (value instanceof String) {
-            return (String) value;
-        } else if (required) {
-            throw new IllegalStateException("Failed to set " + key + " in client data!");
-        }
-        return null;
+    Collection<Breadcrumb> getBreadcrumbs() {
+        return new ArrayList<>(breadcrumbState.getStore());
+    }
+
+    @NonNull
+    AppData getAppData() {
+        return appData;
+    }
+
+    @NonNull
+    DeviceData getDeviceData() {
+        return deviceData;
     }
```

Comment:
Should this now be `getApp` and `getDevice` as per `Event`?

Reply:
These are package-private methods that are only used to access the `DeviceData` object. It's somewhat confusing but `DeviceData` is responsible for generating the `Map` of device information that is put onto the `Event`, and isn't publicly accessible, so we don't need to rename them.

As for why we've added getters I believe previously we were using direct field access, which interfered with mocking in our tests.